k2/csrc/fsa_algo.cu

```diff
@@ -430,6 +430,154 @@ FsaVec LinearFsas(const Ragged<int32_t> &symbols) {
       arcs);
 }
 
+
+FsaVec CtcGraphs(const Ragged<int32_t> &symbols,
+                 Array1<int32_t> *arc_map /*= nullptr*/) {
+  NVTX_RANGE(K2_FUNC);
+  K2_CHECK_EQ(symbols.NumAxes(), 2);
+  ContextPtr &c = symbols.Context();
+
+  int32_t num_fsas = symbols.Dim0();
+  Array1<int32_t> num_states_for(c, num_fsas + 1);
+  int32_t *num_states_for_data = num_states_for.Data();
+  const int32_t *symbol_row_split1_data = symbols.RowSplits(1).Data();
+  // symbols indexed with [fsa][symbol]
+  // for each fsa we need `symbol_num * 2 + 1 + 1` states, `symbol_num * 2 + 1`
+  // means that we need a blank state on each side of a symbol state, `+ 1` is
+  // for final state in k2
+  K2_EVAL(
+      c, num_fsas, lambda_set_num_states, (int32_t fsa_idx0)->void {
+        int32_t symbol_idx0x = symbol_row_split1_data[fsa_idx0],
+                symbol_idx0x_next = symbol_row_split1_data[fsa_idx0 + 1],
+                symbol_num = symbol_idx0x_next - symbol_idx0x;
+        num_states_for_data[fsa_idx0] = symbol_num * 2 + 2;
+      });
+
+  ExclusiveSum(num_states_for, &num_states_for);
+  Array1<int32_t> &fsa_to_states_row_splits = num_states_for;
+  RaggedShape fsa_to_states =
+      RaggedShape2(&fsa_to_states_row_splits, nullptr, -1);
+
+  int32_t num_states = fsa_to_states.NumElements();
+  Array1<int32_t> num_arcs_for(c, num_states + 1);
+  int32_t *num_arcs_for_data = num_arcs_for.Data();
+  const int32_t *fts_row_splits1_data = fsa_to_states.RowSplits(1).Data(),
+                *fts_row_ids1_data = fsa_to_states.RowIds(1).Data(),
+                *symbol_data = symbols.values.Data();
+  // set the arcs number for each state
+  K2_EVAL(
+      c, num_states, lambda_set_num_arcs, (int32_t state_idx01)->void {
+        int32_t fsa_idx0 = fts_row_ids1_data[state_idx01],
+                // we minus fsa_idx0 here, because we are adding one more state,
+                // the final state for each fsa
+                sym_state_idx01 = state_idx01 / 2 - fsa_idx0,
+                remainder = state_idx01 % 2,
+                current_num_arcs = 2;  // normally there are two arcs, self-loop
+                                       // and arc points to the next state
+                                       // blank state always has two arcs
+        if (remainder) {  // symbol state
+          int32_t sym_final_state =
+                    symbol_row_split1_data[fsa_idx0 + 1];
+          // There is no arcs for final states
+          if (sym_state_idx01 == sym_final_state) {
+            current_num_arcs = 0;
+          } else {
+            int32_t current_symbol = symbol_data[sym_state_idx01],
+                    // we set the next symbol of the last symbol to -1, so
+                    // the following if clause will always be true
```

Comment:
It means that the last symbol state would always have 3 arcs. we handle `next_symbol` here, first, to avoid segment fault error, second, to confirm that `current_symbol != next_symbol` so we will assign the last symbol state 3 arcs.

will explain more in the docs.